**레모네이드 판매 예측**


In [ ]:
#라이브러리 사용

import tensorflow as tf
import pandas as pd

In [ ]:
# 데이터 준비
파일경로='https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/lemonade.csv'
데이터=pd.read_csv(파일경로)
데이터.head()

온도  판매량
0  20   40
1  21   42
2  22   44
3  23   46
4  24   48

In [ ]:
# 종속변수, 독립변수
독립= 데이터[['온도']]
종속= 데이터[['판매량']]
print(독립.shape, 종속.shape)

(6, 1) (6, 1)


In [ ]:
# 모델 만들기
X= tf.keras.layers.Input(shape=[1]) # 인풋 레이어 만들기 / 독립변수 col 개수 shape=[1]
Y= tf.keras.layers.Dense(1) (X) # Dense layer / 종속변수 col 개수 Dense(1)
model= tf.keras.models.Model(X,Y)
model.compile(loss='mse')   # 모델이 학습할 방법

In [ ]:
#모델 학습(FIT)
#model.fit(독립, 종속, epochs=10000, verbose=0)  # epochs=학습횟수 verbose=0 화면출력x
model.fit(독립, 종속, epochs=10)  # epochs=학습횟수 verbose=0 화면출력x
# loss: 1.4748e-04 // 1.4748 * 10의 -4승

Epoch 1/10
1/1 [==============================] - 0s 1ms/step - loss: 1.4753e-04
Epoch 2/10
1/1 [==============================] - 0s 949us/step - loss: 1.4766e-04
Epoch 3/10
1/1 [==============================] - 0s 1ms/step - loss: 1.4785e-04
Epoch 4/10
1/1 [==============================] - 0s 2ms/step - loss: 1.4795e-04
Epoch 5/10
1/1 [==============================] - 0s 3ms/step - loss: 1.4798e-04
Epoch 6/10
1/1 [==============================] - 0s 2ms/step - loss: 1.4798e-04
Epoch 7/10
1/1 [==============================] - 0s 2ms/step - loss: 1.4791e-04
Epoch 8/10
1/1 [==============================] - 0s 2ms/step - loss: 1.4783e-04
Epoch 9/10
1/1 [==============================] - 0s 2ms/step - loss: 1.4762e-04
Epoch 10/10
1/1 [==============================] - 0s 2ms/step - loss: 1.4748e-04


In [ ]:
# 모델 이용
model.predict(독립)


array([[40.015057],
       [42.013813],
       [44.012573],
       [46.011333],
       [48.010094],
       [50.008854]], dtype=float32)

In [ ]:
model.predict([[15]])

array([[30.021257]], dtype=float32)

**보스턴 집값 예측**

In [ ]:
# 라이브러리 사용
import pandas as pd
import tensorflow as tf

# 1. 과거의 데이터 준비
path='https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
보스턴=pd.read_csv(path)
print(보스턴.columns)
보스턴.head()

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')


crim    zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07     0  0.469  ...  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18     0  0.458  ...  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18     0  0.458  ...  222     18.7  396.90   5.33  36.2

[5 rows x 14 columns]

In [ ]:
# 독립 , 종속 분리
독립=보스턴[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat']]
종속=보스턴[['medv']]
print(독립.shape, 종속.shape) # 확인 체크 잊지 말기! 버그 적게 할 수 있음

(506, 13) (506, 1)


In [ ]:
# 2. 모델의 구조 만들기
X= tf.keras.layers.Input(shape=[13])
Y=tf.keras.layers.Dense(1)(X)
model=tf.keras.models.Model(X,Y)
model.compile(loss='mse')

In [ ]:
# 3. 데이터로 모델을 학습(fit)
model.fit(독립, 종속, epochs=1000,verbose=0)
model.fit(독립, 종속, epochs=10)

Epoch 1/10
16/16 [==============================] - 0s 924us/step - loss: 25.3227
Epoch 2/10
16/16 [==============================] - 0s 796us/step - loss: 25.3145
Epoch 3/10
16/16 [==============================] - 0s 849us/step - loss: 25.4867
Epoch 4/10
16/16 [==============================] - 0s 1ms/step - loss: 25.4060
Epoch 5/10
16/16 [==============================] - 0s 938us/step - loss: 25.6039
Epoch 6/10
16/16 [==============================] - 0s 1ms/step - loss: 25.6470
Epoch 7/10
16/16 [==============================] - 0s 931us/step - loss: 25.9126
Epoch 8/10
16/16 [==============================] - 0s 941us/step - loss: 25.6657
Epoch 9/10
16/16 [==============================] - 0s 867us/step - loss: 25.6134
Epoch 10/10
16/16 [==============================] - 0s 857us/step - loss: 25.4629


In [ ]:
# 4. 모델 이용
model.predict(독립[0:5]) # 0~5번째 까지 잘라서
model.predict(독립[5:10])

array([[25.65106 ],
       [20.869307],
       [17.750687],
       [ 9.241508],
       [17.309626]], dtype=float32)

In [ ]:
종속[5:10]

medv
5  28.7
6  22.9
7  27.1
8  16.5
9  18.9

In [ ]:
# 모델의 수식 확인
model.get_weights()

**아이리스 품종 분류**

In [ ]:
# 라이브러리 포함
import tensorflow as tf
import pandas as pd

In [ ]:
# 1. 과거 데이터
path='https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
iris=pd.read_csv(path)
iris.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [ ]:
# onehot encoding
인코딩=pd.get_dummies(iris)
인코딩.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0   5.1  3.5    1.4   0.2          1              0             0
1   4.9  3.0    1.4   0.2          1              0             0
2   4.7  3.2    1.3   0.2          1              0             0
3   4.6  3.1    1.5   0.2          1              0             0
4   5.0  3.6    1.4   0.2          1              0             0

In [ ]:
print(인코딩.columns)

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종_setosa', '품종_versicolor',
       '품종_virginica'],
      dtype='object')


In [ ]:
# 독립, 종속
독립= 인코딩[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
종속= 인코딩[['품종_setosa', '품종_versicolor',
       '품종_virginica']]
print(독립.shape, 종속.shape)

(150, 4) (150, 3)


In [ ]:
# 2. 모델의 구조 만들기
X=tf.keras.layers.Input(shape=[4])
Y=tf.keras.layers.Dense(3, activation='softmax')(X)
model=tf.keras.models.Model(X,Y)
model.compile(loss='categorical_crossentropy',metrics='accuracy')

In [ ]:
# FIT
model.fit(독립, 종속, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4441 - accuracy: 0.8400
Epoch 2/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4439 - accuracy: 0.8467
Epoch 3/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4422 - accuracy: 0.8333
Epoch 4/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4404 - accuracy: 0.8333
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4401 - accuracy: 0.8400
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4381 - accuracy: 0.8333
Epoch 7/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4366 - accuracy: 0.8400
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4358 - accuracy: 0.8467
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4363 - accuracy: 0.8333
Epoch 10/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4335 - accuracy: 0.8400


In [ ]:
# model이용
model.predict(독립[-5:])

array([[0.01191692, 0.2253028 , 0.76278025],
       [0.01648557, 0.40937844, 0.574136  ],
       [0.01749022, 0.31189433, 0.67061543],
       [0.00917147, 0.13166983, 0.8591587 ],
       [0.02038972, 0.30082285, 0.6787874 ]], dtype=float32)

In [ ]:
print(종속[-5:])

     품종_setosa  품종_versicolor  품종_virginica
145          0              0             1
146          0              0             1
147          0              0             1
148          0              0             1
149          0              0             1


In [ ]:
# 학습한 가중치 확인
model.get_weights()

[array([[ 0.41939804,  0.4887432 , -0.2369535 ],
        [ 0.3780062 , -0.81450397, -0.02000224],
        [-1.1448007 ,  0.037148  , -0.08544014],
        [-1.1326467 , -0.98399824,  1.0131109 ]], dtype=float32),
 array([ 0.41557106, -0.02004902, -0.27775434], dtype=float32)]

array([[ 0.41939804,  0.4887432 , -0.2369535 ],

[ 0.3780062 , -0.81450397, -0.02000224],

[-1.1448007 ,  0.037148  , -0.08544014],

[-1.1326467 , -0.98399824,  1.0131109 ]],
  dtype=float32),

 array([ 0.41557106, -0.02004902, -0.27775434],
  dtype=float32)
 
 표의 형태라고 생각 ( col: 종속, row: w1~b)

 1col => setosa

 w1=0.41939804

 w2=0.3780062

 w3=-1.1448007

 w4=-1.1326467

 b=0.41557106